AlgebraInLean/Chapter02/Sheet04.lean

```diff
@@ -4,196 +4,209 @@ set_option linter.unusedTactic false
+    rw [←h₁, inv_anticomm]
+    -- END SAMPLE SOLUTION
+    done
 
 /-
+We will now consider another important automorphism called conjugation. For group elements `a` and
+`g`, the conjugation of `a` by `g` is the element `g a g⁻¹`. Two elements `a` and `b` are said to be
+conjugates if there exists some element `g` such that `b = g a g⁻¹`. This is also an equivalence
+relation (like isomorphism), but we will not discuss this now.
+
+You may be familiar with the particular case of invertible n × n matrices (the "general linear
+group" GL(n)), where conjugacy is called matrix similarity (recall that two matrices A and B are
+similar when there is some matrix C such that B = CAC⁻¹).
+-/
 
-As another exercise, let's prove that a specific function mapping within the group of integers
-under addition is a group automorphism.
-
-Specifically, fix G = ⟨ℤ, +⟩, and φ : G → G, x ↦ -x (the function fx = -x).
+variable [Group G]
 
-Note that in order to prove this, we do not necessarily need to "prove" that our φ is an
-endomorphism. We are already defining it as the map φ : G → G (a group onto itself), so it
-suffices to prove that φ is a homomorphism. That may be useful going forward with this proof.
+/-- Conjugation the second argument by the first -/
+def Group.conj : G → G → G := λ (g a : G) ↦ μ (μ g a) (ι g)
 
+/-
+Like `μ`, `conj` is a "curried" function, meaning that rather than taking in two arguments, it takes
+in one argument and returns another function that takes in the second argument. In Lean's syntax,
+this is very convenient to use (`conj g a` is the conjugation of `a` by `g`), and has the benefit
+that we can easily do "partial application" where the function only has one argument present and the
+other is free to vary. That is, `conj g = λ a ↦ g a g⁻¹`. We claim that this partially-applied
+function is an automorphism.
+
+First, here's an easier theorem to get you used to this definition.
 -/
 
-/- A brief definition of our φ: -/
-def φ (x : ℤ) : ℤ := -x
-/- φ : G → G, x ↦ -x -/
+/-- Conjugation by the identity element is just the identity function -/
+theorem Group.conj_id_eq_id : conj (𝕖 : G) = id := by
+  -- SAMPLE SOLUTION
+  unfold conj
+  ext x -- TODO: explain
+  rw [inv_id, id_op, op_id]
+  rfl
+  -- END SAMPLE SOLUTION
+  done
 
-/- Show that φ is a group automorphism -/
-theorem φ_automorphism : ∀ x y : ℤ, φ (x + y) = φ x + φ y ∧ Bijective φ := by
+/-- Conjugation by a particular element g is an automorphism -/
+theorem Group.conj_automorphism (g : G) : Isomorphism (conj g) := by
+  /- Hint: this proof has a lot of rewriting, especially with `op_assoc` -/
   -- SAMPLE SOLUTION
-  intros x y
   constructor
-  /- Prove homomorphism -/
-  · unfold φ
-    rw [neg_add]
-  /- Prove Bijectivity -/
-  · rw [Bijective]
-    constructor
-    /- Injectivity -/
-    · intros x y h
-      unfold φ at h
-      exact neg_inj.mp h
-    /- Surjectivity -/
-    · intro z
-      use -z
-      unfold φ
-      rw [neg_neg]
+  · intro a b
```

Comment:
Also a bit confused with the particular style of this proof. Obviously we have different proofing styles overall...was there something particularly wrong with the original?

Reply:
Sorry just noticed that you changed the original example overall. Did you not want to include the first example about automorphisms? I think it is a lot to throw users immediately into proving that conjugation is an automorphism

Reply:
I replaced that one with the more general one for an arbitrary inverse. I could bring back the negation one if you want (maybe as a warmup for the general one?).

Reply:
As for the proof style stuff, I refactored most of the forward reasoning into backward reasoning. Since this is meant to get them used to how people use Lean properly, and Lean people usually prefer backward reasoning, I think this is helpful. A lot of these proofs lend themselves well to `calc` blocks, but I didn't want to have to explain those on such a short timeline. We could add back more forward reasoning if you want.